### ISSUE TITLE
Differential expression error in pseudo-bulk step

### ISSUE THREAD

**Issue by dfermin** [saezlab/decoupler-py#80]
I am trying to run decoupler on an H5AD file that contains healthy controls (HC) and disease (T2D).

I have successfully created the pseudo-bulk object following the vignette online , specifically this command:

```
pdata = dc.get_pseudobulk(
    adata,
    sample_col='condition',
    groups_col='cell_type',
    layer='counts',
    mode='sum',
    min_cells=10,
    min_counts=1000
)
```

I then subset the data to just 'TAL' cells:
```
tals = pdata[ pdata.obs['celltype'] == 'TAL'].copy()
```

I verify that the object `tal` is an AnnData object with the correct components.
```
AnnData object with n_obs × n_vars = 2 × 22890
    obs: 'condition', 'celltype', 'psbulk_n_cells', 'psbulk_counts'
    var: 'vst.mean', 'vst.variance', 'vst.variance.expected', 'vst.variance.standardized', 'vst.variable'
    layers: 'psbulk_props'
```

Then I run the differential expression command:
```
dds = DeseqDataSet(
    adata=tals,
    design_factors='condition',
    ref_level=['T2D', 'HC'],
    refit_cooks=True,
    n_cpus=1,
)
```

At this point I get the following error:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File ~/mambaforge/envs/decoupler/lib/python3.11/site-packages/pandas/core/indexes/base.py:3790, in Index.get_loc(self, key)
   3789 try:
-> 3790     return self._engine.get_loc(casted_key)
   3791 except KeyError as err:

File index.pyx:152, in pandas._libs.index.IndexEngine.get_loc()

File index.pyx:181, in pandas._libs.index.IndexEngine.get_loc()

File pandas/_libs/hashtable_class_helper.pxi:7080, in pandas._libs.hashtable.PyObjectHashTable.get_item()

File pandas/_libs/hashtable_class_helper.pxi:7088, in pandas._libs.hashtable.PyObjectHashTable.get_item()

KeyError: 'T2D'

The above exception was the direct cause of the following exception:

KeyError                                  Traceback (most recent call last)
Cell In[13], line 1
----> 1 dds = DeseqDataSet(
      2     adata=tals,
      3     design_factors='condition',
      4     ref_level=['T2D', 'HC'],
      5     refit_cooks=True,
      6     n_cpus=1,
      7 )

File ~/mambaforge/envs/decoupler/lib/python3.11/site-packages/pydeseq2/dds.py:257, in DeseqDataSet.__init__(self, adata, counts, metadata, design_factors, continuous_factors, ref_level, min_mu, min_disp, max_disp, refit_cooks, min_replicates, beta_tol, n_cpus, batch_size, joblib_verbosity, quiet)
    253     ref_level = replace_underscores(ref_level)
    255 # Build the design matrix
    256 # Stored in the obsm attribute of the dataset
--> 257 self.obsm["design_matrix"] = build_design_matrix(
    258     metadata=self.obs,
    259     design_factors=self.design_factors,
    260     continuous_factors=self.continuous_factors,
    261     ref_level=ref_level,
    262     expanded=False,
    263     intercept=True,
    264 )
    266 # Check that the design matrix has full rank
    267 self._check_full_rank_design()

File ~/mambaforge/envs/decoupler/lib/python3.11/site-packages/pydeseq2/utils.py:231, in build_design_matrix(metadata, design_factors, ref_level, continuous_factors, expanded, intercept)
    229 if len(ref_level) != 2:
    230     raise KeyError("The reference level should contain 2 strings.")
--> 231 if ref_level[1] not in metadata[ref_level[0]].values:
    232     raise KeyError(
    233         f"The metadata data should contain a '{ref_level[0]}' column"
    234         f" with a '{ref_level[1]}' level."
    235     )
    237 # Check that the reference level is not in the matrix (if unexpanded design)

File ~/mambaforge/envs/decoupler/lib/python3.11/site-packages/pandas/core/frame.py:3896, in DataFrame.__getitem__(self, key)
   3894 if self.columns.nlevels > 1:
   3895     return self._getitem_multilevel(key)
-> 3896 indexer = self.columns.get_loc(key)
   3897 if is_integer(indexer):
   3898     indexer = [indexer]

File ~/mambaforge/envs/decoupler/lib/python3.11/site-packages/pandas/core/indexes/base.py:3797, in Index.get_loc(self, key)
   3792     if isinstance(casted_key, slice) or (
   3793         isinstance(casted_key, abc.Iterable)
   3794         and any(isinstance(x, slice) for x in casted_key)
   3795     ):
   3796         raise InvalidIndexError(key)
-> 3797     raise KeyError(key) from err
   3798 except TypeError:
   3799     # If we have a listlike key, _check_indexing_error will raise
   3800     #  InvalidIndexError. Otherwise we fall through and re-raise
   3801     #  the TypeError.
   3802     self._check_indexing_error(key)

KeyError: 'T2D'
```

Any suggestions for how to fix this?
It might not be a bug but a as far as I can tell my input AnnData is correctly formatted.

I'm doing this in a Conda/Mamba environment on an Apple computer. The environment has the following packages installed:
```
Package                       Version
----------------------------- ------------
aioeasywebdav                 2.4.0
aiohttp                       3.8.5
aiosignal                     1.3.1
amply                         0.1.6
anndata                       0.9.2
annotated-types               0.5.0
anyio                         4.0.0
appdirs                       1.4.4
argon2-cffi                   23.1.0
argon2-cffi-bindings          21.2.0
arrow                         1.2.3
asttokens                     2.4.0
async-timeout                 4.0.3
attmap                        0.13.2
attrs                         23.1.0
backcall                      0.2.0
backports.functools-lru-cache 1.6.5
bcrypt                        4.0.1
beautifulsoup4                4.12.2
bleach                        6.0.0
boto3                         1.28.53
botocore                      1.31.53
Brotli                        1.1.0
cached-property               1.5.2
cachetools                    5.3.1
certifi                       2023.7.22
cffi                          1.15.1
charset-normalizer            3.2.0
colorama                      0.4.6
comm                          0.1.4
ConfigArgParse                1.7
connection-pool               0.0.3
contourpy                     1.1.1
cycler                        0.11.0
datrie                        0.8.2
debugpy                       1.8.0
decorator                     5.1.1
decoupler                     1.5.0
defusedxml                    0.7.1
docrep                        0.3.2
docutils                      0.20.1
dpath                         2.1.6
dropbox                       11.36.2
eido                          0.2.1
entrypoints                   0.4
exceptiongroup                1.1.3
executing                     1.2.0
fastjsonschema                2.18.0
filechunkio                   1.8
fonttools                     4.39.4
fqdn                          1.5.1
frozenlist                    1.4.0
ftputil                       5.0.4
gitdb                         4.0.10
GitPython                     3.1.37
google-api-core               2.11.1
google-api-python-client      2.100.0
google-auth                   2.23.0
google-auth-httplib2          0.1.1
google-cloud-core             2.3.3
google-cloud-storage          2.11.0
google-crc32c                 1.1.2
google-resumable-media        2.6.0
googleapis-common-protos      1.60.0
grpcio                        1.57.0
h5py                          3.8.0
httplib2                      0.22.0
humanfriendly                 10.0
idna                          3.4
importlib-metadata            6.8.0
importlib-resources           6.1.0
inflect                       7.0.0
iniconfig                     2.0.0
ipykernel                     6.25.2
ipython                       8.15.0
ipython-genutils              0.2.0
ipywidgets                    8.1.1
isoduration                   20.11.0
jedi                          0.19.0
Jinja2                        3.1.2
jmespath                      1.0.1
joblib                        1.3.2
jsonpointer                   2.4
jsonschema                    4.19.1
jsonschema-specifications     2023.7.1
jupyter                       1.0.0
jupyter_client                8.3.1
jupyter-console               6.6.3
jupyter_core                  5.3.1
jupyter-events                0.7.0
jupyter_server                2.7.3
jupyter_server_terminals      0.4.4
jupyterlab-pygments           0.2.2
jupyterlab-widgets            3.0.9
kiwisolver                    1.4.5
llvmlite                      0.40.0
logmuse                       0.2.6
markdown-it-py                3.0.0
MarkupSafe                    2.1.3
matplotlib                    3.8.0
matplotlib-inline             0.1.6
mdurl                         0.1.0
mistune                       3.0.1
multidict                     6.0.4
natsort                       8.4.0
nbclassic                     1.0.0
nbclient                      0.8.0
nbconvert                     7.8.0
nbformat                      5.9.2
nest-asyncio                  1.5.6
networkx                      3.1
notebook                      6.5.4
notebook_shim                 0.2.3
numba                         0.57.0
numpy                         1.24.0
oauth2client                  4.1.3
omnipath                      1.0.7
overrides                     7.4.0
packaging                     23.1
pandas                        2.1.1
pandocfilters                 1.5.0
paramiko                      3.3.1
parso                         0.8.3
patsy                         0.5.3
peppy                         0.35.7
pexpect                       4.8.0
pickleshare                   0.7.5
Pillow                        9.5.0
pip                           23.2.1
pkgutil_resolve_name          1.3.10
plac                          1.4.0
platformdirs                  3.10.0
pluggy                        1.3.0
ply                           3.11
prettytable                   3.9.0
prometheus-client             0.17.1
prompt-toolkit                3.0.39
protobuf                      4.23.4
psutil                        5.9.5
ptyprocess                    0.7.0
PuLP                          2.7.0
pure-eval                     0.2.2
pyasn1                        0.5.0
pyasn1-modules                0.3.0
pycparser                     2.21
pydantic                      2.0.3
pydantic_core                 2.3.0
pydeseq2                      0.4.0
Pygments                      2.16.1
PyNaCl                        1.5.0
pynndescent                   0.5.10
pyOpenSSL                     23.2.0
pyparsing                     3.1.1                                                                                                                                                                                [30/366]
PyQt5                         5.15.9
PyQt5-sip                     12.12.2
pysftp                        0.2.9
PySocks                       1.7.1
pytest                        7.4.2
python-dateutil               2.8.2
python-irodsclient            1.1.8
python-json-logger            2.0.7
pytz                          2023.3.post1
pyu2f                         0.1.5
PyYAML                        6.0.1
pyzmq                         25.1.1
qtconsole                     5.4.4
QtPy                          2.4.0
referencing                   0.30.2
requests                      2.31.0
reretry                       0.11.8
rfc3339-validator             0.1.4
rfc3986-validator             0.1.1
rich                          13.5.3
rpds-py                       0.10.3
rsa                           4.9
s3transfer                    0.6.2
scanpy                        1.9.3
scikit-learn                  1.2.2
scipy                         1.11.2
seaborn                       0.12.2
Send2Trash                    1.8.2
session-info                  1.0.0
setuptools                    68.2.2
setuptools-scm                7.1.0
sip                           6.7.11
six                           1.16.0
slacker                       0.14.0
smart-open                    6.4.0
smmap                         3.0.5
snakemake                     7.32.4
sniffio                       1.3.0
soupsieve                     2.5
stack-data                    0.6.2
statsmodels                   0.14.0
stdlib-list                   0.8.0
stone                         3.3.1
stopit                        1.1.2
tabulate                      0.9.0
terminado                     0.17.1
threadpoolctl                 3.1.0
throttler                     1.2.2
tinycss2                      1.2.1
toml                          0.10.2
tomli                         2.0.1
toposort                      1.10
tornado                       6.3.3
tqdm                          4.66.1
traitlets                     5.9.0
typing_extensions             4.8.0
typing-utils                  0.1.0
tzdata                        2023.3
ubiquerg                      0.6.3
umap-learn                    0.5.3
uri-template                  1.3.0
uritemplate                   4.1.1
urllib3                       1.26.16
veracitools                   0.1.3
wcwidth                       0.2.6
webcolors                     1.13
webencodings                  0.5.1
websocket-client              1.6.3
wheel                         0.41.2
widgetsnbextension            4.0.9
wrapt                         1.15.0
yarl                          1.9.2
yte                           1.5.1
zipp                          3.17.0
```

**Comment by PauBadiaM** [comment]
Hi @dfermin,

Looking at the `tals` object I see that there are only 2 samples in your filtered pseudobulk object. What I thing might be happening is that you don't have any sample for `T2D` that passes the pseudobulking thresholds and you end up with only two samples of `HC`. Check the `.obs` attribute to confirm this. Unfortunately, to run any statistical test you need at least 2 samples per group (2 `T2D` and 2 `HC` for example), therefore it might be the case that such contrast is not possible with your data. Hope this is helpful!

**Comment by demian1** [comment]
Hi @dfermin,

I was taking a look again at this issue. It seems that you used the `'condition'` column as the id for the column which should be sample ids (i.e. the `sample_col` parameter) I think what you ended up doing, is sum all the counts of a cell type across all samples.

Please make sure that you have something similar to this instead:

```
pdata = dc.get_pseudobulk(
    adata,
    sample_col='patient', # or 'sample', whatever it is in your .obs column
    groups_col='cell_type',
    layer='counts',
    mode='sum',
    min_cells=10,
    min_counts=1000
)
```

Hope this was helpful

**Comment by dfermin** [comment]
Hi Pau

Sorry for the late response.
I've updated the 'obs' object to contain the sample IDs as a column. 
However even after I recreate the psuedobulk object I still get the same error as before. 
```
## add sample IDs to the obs data.frame
adata.obs['sample'] = adata.obs.index
adata.obs['sample'] = adata.obs['sample'].apply(lambda x: x.split("_")[0])
```
```
## create pseudobulk object by sample
pdata = dc.get_pseudobulk(
    adata,
    sample_col='sample', 
    groups_col='cell_type',
    layer='counts',
    mode='sum',
    min_cells=10,
    min_counts=100
)

## subset to just 'tals' cells
tals = pdata[ pdata.obs['celltype'] == 'TAL'].copy()

## run Deseq
dds = DeseqDataSet(
    adata=tals,
    design_factors='condition',
    ref_level=['T2D', 'HC'],
    refit_cooks=True,
    n_cpus=1,
)
```

A quick look at tals.obs tells me we have both 'T2D' and 'HC' samples:
```
tals.obs.groupby('condition').count()['sample']
condition
HC     6
T2D    6
Name: sample, dtype: int64
```

**Comment by dfermin** [comment]
Hi Pau

I wanted to close this issue since I have overcome it.
So I was able to run the code successfully after dropping the `ref_level` argument to the `DeseqDataSet` function.
I had included that argument as per the vignette for decoupleR but it's removal was the only way the function would run. 
I don't thin this is a decoupleR bug but something todo with either Deseq or the vignette needs updating(?)
Either way I'm closing this issue. 

Thanks for your help!
